### PR TITLE
test: improve the test for SpoonTagger class

### PR DESCRIPTION
**Pull request by Rohitesh-Kumar-Jain**
Hi, I have improved the mutation strength from 0% to 100% for the SpoonTagger class.

Here's the report before improvement.

Screen Shot 2021-06-08 at 22 20 25-fullpage

Actually, the test was poorly written, it had overlooked the scenario when the Spoon.java file already exists. I have taken care of that scenario and have manually tested the mutation "Removed all instructions in the method", and it is getting killed now.

I feel the PR is ready for merging.

### CODE REVIEW

**Comment by Rohitesh-Kumar-Jain** [comment]
#1818
Link to #3967

**Comment by slarse** [comment]
Hi @Rohitesh-Kumar-Jain,

Just FYI I will get back to reviewing your PRs tomorrow, was a bit busy with other things today :)

**Comment by Rohitesh-Kumar-Jain** [comment]
> reviewing your PRs tomorrow, was a bit busy with other things today :)

Yeah no problem

File: src/test/java/spoon/processing/ProcessingTest.java [hunk 1]
```diff
@@ -87,8 +87,13 @@ public void testRuntimeProcessorManager() {
 
 	@Test
 	public void testSpoonTagger() {
+		// contract: this tests the process method of the SpoonTagger class.
```
Comment (slarse) [inline src/test/java/spoon/processing/ProcessingTest.java:90]: Like I mentioned in the other review, a contract should state what is expected to happen given what the test does. An appropriate contract here would be: `contract: after running SpoonTagger, the file spoon/Spoon.java should exist in the output directory`

File: src/test/java/spoon/processing/ProcessingTest.java [hunk 2]
```diff
@@ -87,8 +87,13 @@ public void testRuntimeProcessorManager() {
...
 		final Launcher launcher = new Launcher();
 		launcher.addProcessor("spoon.processing.SpoonTagger");
+		// we need to make sure that the spoon/Spoon.java file doesn't already exist in the system, as it will cause the assertion to be irrelevant
```
Comment (slarse) [inline src/test/java/spoon/processing/ProcessingTest.java:93]: This is an example of a good comment: it says _why_ we do something.

File: src/test/java/spoon/processing/ProcessingTest.java [hunk 3]
```diff
@@ -87,8 +87,13 @@ public void testRuntimeProcessorManager() {
...
+		File file = new File (launcher.getModelBuilder().getSourceOutputDirectory() + "/spoon/Spoon.java");
+		// deleting the Spoon.java if it exists
```
Comment (slarse) [inline src/test/java/spoon/processing/ProcessingTest.java:95]: This is an example of a bad comment: it just restates _what_ the code does. The code already says that.

**Review by slarse — CHANGES_REQUESTED** [review]
This is a good find!

For tests that require that require a "clean slate" when running, it's better to use a temporary directory instead of deleting select files. You can easily create a temporary directory with JUnit5 like so:

@Test
public void testSpoonTagger(@TempDir File tempDir) {
    // contract: ...
    Launcher launcher = new Launcher();
    launcher.setSourceOutputDirectory(tempDir.getAbsolutePath());
    // do the test here
}
```

**Comment by Rohitesh-Kumar-Jain** [comment]
> For tests that require that require a "clean slate" when running, it's better to use a temporary directory instead of deleting select files. 

Thanks for letting me know this, I will keep this in my mind for future tests

File: src/test/java/spoon/processing/ProcessingTest.java (hunk 1 above)
Comment (Rohitesh-Kumar-Jain) [reply src/test/java/spoon/processing/ProcessingTest.java:90]: Thanks for correcting the contract, I will make sure that future test contracts are more appropriate.

File: src/test/java/spoon/processing/ProcessingTest.java (hunk 3 above)
Comment (Rohitesh-Kumar-Jain) [reply src/test/java/spoon/processing/ProcessingTest.java:95]: Thanks for pointing out, I will avoid redundant comments in future

**Comment by slarse** [comment]
@Rohitesh-Kumar-Jain Given your responses I'm not sure it was clear, but my comments here were not meant only for future PRs, but for this one as well! You can see it as "changes requested" on the PR review. Please address the comments :)

**Comment by Rohitesh-Kumar-Jain** [comment]
> @Rohitesh-Kumar-Jain Given your responses I'm not sure it was clear, but my comments here were not meant only for future PRs, but for this one as well! You can see it as "changes requested" on the PR review. Please address the comments :)

Ik this is meant for this PR only, I was saying that I will keep these things in my mind for future PRs.

**Comment by slarse** [comment]
> I was saying that I will keep these things in my mind for future PRs.

I understood that. What I did not know for sure was whether you understood that I meant for the comments to be fixed in this PR, and not only as suggestions for future PRs, and so I clarified the fact.

**Comment by slarse** [comment]
@Rohitesh-Kumar-Jain This looks fine to me, are you done here?

**Comment by Rohitesh-Kumar-Jain** [comment]
> @Rohitesh-Kumar-Jain This good to me, are you done here?

Yes, this was a small class and I believe all 3 Descartes parameters have been improved to 100% 🎉

File: src/test/java/spoon/processing/ProcessingTest.java (hunk 3 above)
Comment (slarse) [reply src/test/java/spoon/processing/ProcessingTest.java]: Note that as with any rule of thumb, there are exceptions. Sometimes we simply fail to write code that is clear enough to be understood, and then we need to put a comment in to clarify. That wasn't the case here though, the code was very clear.

**Review by slarse — APPROVED** [review]
@Rohitesh-Kumar-Jain 
> Yes, this was a small class and I believe all 3 Descartes parameters have been improved to 100%

That's excellent!

@monperrus ping for merge, one pseudo-tested method defeated!

File: src/test/java/spoon/processing/ProcessingTest.java [hunk 4]
```diff
@@ -86,8 +87,10 @@ public void testRuntimeProcessorManager() {
 	}
 
 	@Test
-	public void testSpoonTagger() {
+	public void testSpoonTagger(@TempDir File tempDir) {
+		// contract: after running SpoonTagger, the file spoon/Spoon.java should exist in the output directory
```
Comment (slarse) [inline src/test/java/spoon/processing/ProcessingTest.java:91]: @Rohitesh-Kumar-Jain By the way, this is a well specified contract. Nice.

**Comment by monperrus** [comment]
> one pseudo-tested method defeated!

Hurray, that's an achievement, per the plan pf GSOC!

@Rohitesh-Kumar-Jain thanks a lot for your patience and determination to understand what a good test is and how to improve Spoon's test suite.

@slarse thanks a lot for the mentoring

**Pull request merged**